src/webgpu/api/operation/render_pipeline/sample_mask.spec.ts

```diff
@@ -1,16 +1,443 @@
+    });
+    const resolveTargetTexture =
+      sampleCount === 1
+        ? null
+        : this.device.createTexture({
+            format,
+            size: {
+              width: kRenderTargetSize,
+              height: kRenderTargetSize,
+              depthOrArrayLayers: 1,
+            },
+            sampleCount: 1,
+            mipLevelCount: 1,
+            usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.RENDER_ATTACHMENT,
+          });
+
+    const renderPassDescriptor: GPURenderPassDescriptor = {
+      colorAttachments: [
+        {
+          view: renderTargetTexture.createView(),
+          resolveTarget: resolveTargetTexture?.createView(),
+
+          clearValue: { r: 0.0, g: 0.0, b: 0.0, a: 0.0 },
+          loadOp: 'clear',
+          storeOp: 'store',
+        },
+      ],
+    };
+    const commandEncoder = this.device.createCommandEncoder();
+    const passEncoder = commandEncoder.beginRenderPass(renderPassDescriptor);
+    passEncoder.setPipeline(pipeline);
+    passEncoder.setBindGroup(0, uniformBindGroup);
+
+    if (rasterizationMask === 15) {
+      // draw full screen quad
+      passEncoder.draw(6);
+    } else {
+      if ((rasterizationMask & 1) !== 0) {
+        // draw top-left quad
+        passEncoder.draw(6, 1, 6);
+      }
+      if ((rasterizationMask & 2) !== 0) {
+        // draw top-right quad
+        passEncoder.draw(6, 1, 12);
+      }
+      if ((rasterizationMask & 4) !== 0) {
+        // draw bottom-left quad
+        passEncoder.draw(6, 1, 18);
+      }
+      if ((rasterizationMask & 8) !== 0) {
+        // draw bottom-right quad
+        passEncoder.draw(6, 1, 24);
+      }
+    }
+    passEncoder.end();
+    this.device.queue.submit([commandEncoder.finish()]);
+
+    const result = sampleCount === 1 ? renderTargetTexture : resolveTargetTexture;
+    assert(result !== null);
+    return result;
+  }
+
+  CheckSingleSampledResult(
+    texture: GPUTexture,
+    rasterizationMask: number,
+    sampleMask: number,
+    fragmentShaderOutputMask: number
+  ) {
+    let expected: typeof kColors[number] = kEmptySample;
+
+    if (
+      // rasterization needs to cover the center of the pixel
+      (rasterizationMask & 15) >= 15 &&
+      // There is only one sample
+      (fragmentShaderOutputMask & sampleMask & 1) !== 0
+    ) {
+      // When full screen quad is drawn, Texel 3 is sampled at the pixel center
+      expected = kColors[3];
+    }
+
+    this.expectSingleColor(texture, format, {
+      size: [1, 1, 1],
+      exp: {
+        R: expected.R / 0xff,
+        G: expected.G / 0xff,
+        B: expected.B / 0xff,
+        A: expected.A / 0xff,
+      },
+    });
+  }
+
+  CheckMultiSampledResult(
+    texture: GPUTexture,
+    sampleCount: number,
+    rasterizationMask: number,
+    sampleMask: number,
+    fragmentShaderOutputMask: number
+  ) {
+    const expected = {
+      R: 0,
+      G: 0,
+      B: 0,
+      A: 0,
+    };
+    range(4, (id: number) => {
```

Comment:
nit: I'd write this as a (C-style) `for` loop, it's a little hard to read that this is a loop.